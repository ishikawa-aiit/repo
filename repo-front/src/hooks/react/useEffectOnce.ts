import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
