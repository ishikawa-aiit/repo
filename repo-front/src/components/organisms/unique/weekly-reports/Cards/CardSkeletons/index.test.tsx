import { render, screen, within } from '@testing-library/react'

import { CardSkeletons } from '.'

describe('Card', () => {
  it('正常にレンダリングされている', () => {
    const { container } = render(<CardSkeletons />)
    const { getAllByRole } = within(screen.getByTestId('CardSkeletons'))

    expect(getAllByRole('status')).toHaveLength(3)
    expect(container).toMatchSnapshot()
  })
})
