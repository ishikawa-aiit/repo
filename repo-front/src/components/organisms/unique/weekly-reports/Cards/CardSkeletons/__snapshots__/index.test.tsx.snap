// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 6px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 3px;
  padding-bottom: 3px;
  border: 1px;
  border-color: neutral.200;
  border-radius: 12px;
  width: 100%;
  background-color: bg.white;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.emotion-2>*:not(style)~*:not(style) {
  margin-top: 3px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-3 {
  box-shadow: none;
  -webkit-background-clip: padding-box;
  background-clip: padding-box;
  cursor: default;
  color: transparent;
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 8px;
}

.emotion-3::before,
.emotion-3::after,
.emotion-3 * {
  visibility: hidden;
}

<div>
  <div
    class="chakra-stack emotion-0"
    data-cy="CardSkeletons"
    data-testid="CardSkeletons"
    role="list"
  >
    <div
      aria-busy="true"
      class="emotion-1"
      role="status"
    >
      <div
        class="chakra-stack emotion-2"
      >
        <div
          class="chakra-skeleton emotion-3"
        />
        <div
          class="chakra-skeleton emotion-3"
        />
        <div
          class="chakra-skeleton emotion-3"
        />
      </div>
    </div>
    <div
      aria-busy="true"
      class="emotion-1"
      role="status"
    >
      <div
        class="chakra-stack emotion-2"
      >
        <div
          class="chakra-skeleton emotion-3"
        />
        <div
          class="chakra-skeleton emotion-3"
        />
        <div
          class="chakra-skeleton emotion-3"
        />
      </div>
    </div>
    <div
      aria-busy="true"
      class="emotion-1"
      role="status"
    >
      <div
        class="chakra-stack emotion-2"
      >
        <div
          class="chakra-skeleton emotion-3"
        />
        <div
          class="chakra-skeleton emotion-3"
        />
        <div
          class="chakra-skeleton emotion-3"
        />
      </div>
    </div>
  </div>
</div>
`;
